pkg/engine/lifecycletest/error_hooks_test.go

```diff
@@ -0,0 +1,2018 @@
+		require.Error(t, err)
+		require.ErrorContains(t, err, "maximum number of error hook retries reached")
+
+		sawWarning := false
+		for _, evt := range evts {
+			if evt.Type != DiagEvent {
+				continue
+			}
+			d := evt.Payload().(DiagEventPayload)
+			if d.Severity == diag.Warning && strings.Contains(d.Message, "maximum number of error hook retries") {
+				sawWarning = true
+				break
+			}
+		}
+		require.True(t, sawWarning, "expected a warning diagnostic when retry limit is hit")
+		return err
+	}
+
+	project := p.GetProject()
+	_, err := lt.TestOp(Update).RunStep(project, p.GetTarget(t, nil), p.Options, false, p.BackendClient, validateWarn, "0")
+	require.Error(t, err)
+}
+
+func TestErrorHooks_RetryLimitWarningAt100_Update(t *testing.T) {
+	t.Parallel()
+
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{
+				CreateF: func(_ context.Context, req plugin.CreateRequest) (plugin.CreateResponse, error) {
+					return plugin.CreateResponse{ID: "id", Properties: resource.PropertyMap{}, Status: resource.StatusOK}, nil
+				},
+				UpdateF: func(_ context.Context, req plugin.UpdateRequest) (plugin.UpdateResponse, error) {
+					if req.Preview {
+						return plugin.UpdateResponse{Status: resource.StatusOK}, nil
+					}
+					return plugin.UpdateResponse{Status: resource.StatusPartialFailure}, errors.New("update failed")
+				},
+			}, nil
+		}),
+	}
+
+	isUpdate := false
+
+	programF := deploytest.NewLanguageRuntimeF(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		callbacks, err := deploytest.NewCallbacksServer()
+		require.NoError(t, err)
+		defer func() { require.NoError(t, callbacks.Close()) }()
+
+		h, err := deploytest.NewHook(monitor, callbacks, "hook",
+			func(_ context.Context, _ resource.URN, _ resource.ID, _ string, _ tokens.Type,
+				_, _, _, _ resource.PropertyMap, _ string, errs []string,
+			) (bool, error) {
+				require.NotEmpty(t, errs)
+				return true, nil
+			}, true)
+		require.NoError(t, err)
+
+		inputs := resource.NewPropertyMapFromMap(map[string]any{"v": "a"})
+		if isUpdate {
+			inputs = resource.NewPropertyMapFromMap(map[string]any{"v": "b"})
+		}
+
+		_, err = monitor.RegisterResource("pkgA:m:typA", "resA", true, deploytest.ResourceOptions{
+			Inputs: inputs,
+			ResourceHookBindings: deploytest.ResourceHookBindings{
+				OnError: []*deploytest.ResourceHook{h},
+			},
+		})
+		require.NoError(t, err)
+
+		err = monitor.SignalAndWaitForShutdown(context.Background())
+		require.NoError(t, err)
+		return nil
+	})
+
+	hostF := deploytest.NewPluginHostF(nil, nil, programF, loaders...)
+	p := &lt.TestPlan{
+		Options: lt.TestUpdateOptions{T: t, HostF: hostF},
+	}
+	validateCreate := func(_ workspace.Project, _ deploy.Target, _ JournalEntries, _ []Event, err error) error {
+		require.NoError(t, err)
+		isUpdate = true
```

Comment:
If ind it a little weird that we set this in the validation function. Can we set `isUpdate = true` before calling `lt.TestOp(Update).RunStep` below and get rid of this `validateCreate` function?